crates/extensions/tedge_mqtt_ext/src/lib.rs

```diff
@@ -187,36 +234,30 @@ impl Actor for MqttActor {
     }
 
     async fn run(&mut self) -> Result<(), RuntimeError> {
+        let (mut inbox, outbox) = self
+            .messages
+            .take()
+            .expect("MQTT actor can only be run once")
+            .split();
+
         let mut mqtt_client = tokio::select! {
             connection = mqtt_channel::Connection::new(&self.mqtt_config) => {
                 connection.map_err(Box::new)?
             }
-            Some(RuntimeRequest::Shutdown) = self.messages.recv_signal() => {
+            Some(RuntimeRequest::Shutdown) = inbox.recv_signal() => {
                 // Shutdown requested even before the connection has been established
                 return Ok(())
             }
         };
 
-        loop {
-            tokio::select! {
-                message_or_signal = self.messages.try_recv() => {
-                    match message_or_signal {
-                        Ok(Some(message)) => {
-                            mqtt_client
-                                .published
-                                .send(message)
-                                .await
-                                .map_err(Box::new)?
-                        }
-                        Ok(None) | Err(RuntimeRequest::Shutdown) => break,
-                    }
-                }
-                Some(message) = mqtt_client.received.next() => {
-                    self.messages.send(message).await?
-                },
-                else => break,
-            }
+        pin! {
+            let send_mqtt = inbox.forward_to_mqtt(&mut mqtt_client.published);
+            let receive_mqtt = outbox.forward_to_actors(&mut mqtt_client.received);
         }
-        Ok(())
+
+        futures::future::select(send_mqtt, receive_mqtt)
```

Comment:
Yeah, I think we can get rid of the pinning requirement by creating our own helper function. The reason why select needs the futures to be pinned is because it returns the non completed future, but we don't care in this case.